# Lab for taxonomic annotation using different databases and tools

We talked about how profoundly our choice of database (or lack of reference sequences) can impact our ability to interpret our community sequencing datasets. Now, you all are going to see for yourself how significant this impact can be!

We will use the following tools:
- `sourmash`
- `EUKulele`

And look at the following databases:
- NCBI `GenBank` database for protozoa
- Marine Microbial Eukaryote Transcriptome Sequencing Project (MMETSP) database with MarRef bacterial sequences
- Marine Microbial Eukaryote Transcriptome Sequencing Project (MMETSP) databas
e with MarRef bacterial sequences and _Phaeocystis_ colonial transcriptomes
- Marine Microbial Eukaryote Transcriptome Sequencing Project (MMETSP) databas
e with MarRef bacterial sequences and _Phaeocystis_ free-living transcriptomes
- Marine Microbial Eukaryote Transcriptome Sequencing Project (MMETSP) databas
e with MarRef bacterial sequences and all available _Phaeocystis_ transcriptomes

## Group assignments:

Take a look at both `.csv` files in the folders in:

1. Group 1 = `output/taxonomy_annotation_workflow/EUKulele_Database1/taxonomy_estimation`
2. Group 2 = `output/taxonomy_annotation_workflow/EUKulele_Database3/taxonomy_estimation`
2. Group 3 = `output/taxonomy_annotation_workflow/EUKulele_Database4/taxonomy_estimation`
