#!/bin/sh
# properties = {"type": "single", "rule": "mmseqs_taxonomy", "local": false, "input": ["../../output/taxonomy_annotation_workflow/subsampled_assemblies/large.fasta"], "output": ["../../output/taxonomy_annotation_workflow/mmseqs_UniRef100/large.tsv"], "wildcards": {"database": "UniRef100", "size_fraction": "large"}, "params": {"mmseqs_database": "/vortexfs1/omics/alexander/data/databases/mmseqs2/UniRef100", "query_db": "../../output/taxonomy_annotation_workflow/mmseqs_UniRef100/large_query", "tmp_mmseqs": "../../output/taxonomy_annotation_workflow/mmseqs_UniRef100/large_tmp", "mmseqs_result_folder": "../../output/taxonomy_annotation_workflow/mmseqs_UniRef100/large"}, "log": [], "threads": 1, "resources": {"mem_mb": 1000, "mem_mib": 954, "disk_mb": 1000, "disk_mib": 954, "tmpdir": "<TBD>"}, "jobid": 6, "cluster": {"account": "akrinos", "command_options": {"slurm": {"account": "--account={}", "command": "sbatch --parsable --qos=unlim", "key_mapping": null, "mem": "--mem={}g", "name": "--job-name={}", "nodes": "-N {}", "queue": "--partition={}", "threads": "-n {}", "time": "--time={}"}}, "cpupertask": 18, "mem": 180, "nodes": 1, "queue": "scavenger", "system": "slurm", "tasks": 1, "time": 5000}}
cd /vortexfs1/omics/alexander/akrinos/2023-env-bioinformatics/Lab_Taxonomy/code/snakemake-workflow && /vortexfs1/home/akrinos/.conda/envs/ssub22/envs/snakemake7/bin/python3.11 -m snakemake --snakefile '/vortexfs1/omics/alexander/akrinos/2023-env-bioinformatics/Lab_Taxonomy/code/snakemake-workflow/Snakefile' --target-jobs 'mmseqs_taxonomy:database=UniRef100,size_fraction=large' --allowed-rules 'mmseqs_taxonomy' --cores 'all' --attempt 1 --force-use-threads  --resources 'mem_mb=1000' 'mem_mib=954' 'disk_mb=1000' 'disk_mib=954' --wait-for-files '/vortexfs1/omics/alexander/akrinos/2023-env-bioinformatics/Lab_Taxonomy/code/snakemake-workflow/.snakemake/tmp.5sflhcfn' '../../output/taxonomy_annotation_workflow/subsampled_assemblies/large.fasta' --force --keep-target-files --keep-remote --max-inventory-time 0 --nocolor --notemp --no-hooks --nolock --ignore-incomplete --rerun-triggers 'software-env' 'code' 'params' 'mtime' 'input' --skip-script-cleanup  --use-conda  --conda-frontend 'mamba' --conda-base-path '/vortexfs1/home/akrinos/.conda/envs/ssub22/envs/snakemake7' --wrapper-prefix 'https://github.com/snakemake/snakemake-wrappers/raw/' --latency-wait 5 --scheduler 'ilp' --scheduler-solver-path '/vortexfs1/home/akrinos/.conda/envs/ssub22/envs/snakemake7/bin' --default-resources 'mem_mb=max(2*input.size_mb, 1000)' 'disk_mb=max(2*input.size_mb, 1000)' 'tmpdir=system_tmpdir' --mode 2 && touch '/vortexfs1/omics/alexander/akrinos/2023-env-bioinformatics/Lab_Taxonomy/code/snakemake-workflow/.snakemake/tmp.5sflhcfn/6.jobfinished' || (touch '/vortexfs1/omics/alexander/akrinos/2023-env-bioinformatics/Lab_Taxonomy/code/snakemake-workflow/.snakemake/tmp.5sflhcfn/6.jobfailed'; exit 1)

